### PR TITLE
URL previewing support

### PR DESCRIPTION
**Pull request by ara4n**
Adds:
 * Ability for ImageView to be used to preview generic image URLs taken from URL metadata
 * Ability to un-hide hidden image previews from MessageContextMenu (updating localStorage and the event's widgetHidden property)
 * CSS for LinkPreviewWidget

This is part of a set of PRs spanning vector-web, matrix-react-sdk, matrix-js-sdk and synapse.

### CODE REVIEW

File: src/components/views/elements/ImageView.js [hunk 1]
```diff
@@ -27,6 +27,12 @@ module.exports = React.createClass({
     displayName: 'ImageView',
 
     propTypes: {
+        mxEvent: React.PropTypes.object,
```
Comment (richvdh) [inline src/components/views/elements/ImageView.js]: some comments would be nice...

File: src/components/views/elements/ImageView.js [hunk 2]
```diff
@@ -27,6 +27,12 @@ module.exports = React.createClass({
...
+        src: React.PropTypes.string.isRequired,
+        link: React.PropTypes.string,
+        width: React.PropTypes.number,
+        height: React.PropTypes.number,
+        size: React.PropTypes.number,
```
Comment (richvdh) [inline src/components/views/elements/ImageView.js]: `size` is a horribly overloaded term in relation to images. `fileSize`? `byteCount`?

File: src/components/views/elements/ImageView.js [hunk 3]
```diff
@@ -72,11 +78,15 @@ module.exports = React.createClass({
         if(this.props.name) {
             name = this.props.name;
         } else if(this.props.mxEvent) {
-            name = props.mxEvent.getContent().body;
+            name = this.props.mxEvent.getContent().body;
```
Comment (richvdh) [inline src/components/views/elements/ImageView.js]: I feel like this component has no business messing with `mxEvent`, and the poking in `mxEvent` should be left to the parent. I'd understand if you didn't fancy messing with this right now, but let's not propagate the evil. I'd rather see this line removed than fixed.

File: src/components/views/rooms/MessageContextMenu.js [hunk 4]
```diff
@@ -66,6 +66,16 @@ module.exports = React.createClass({
         if (this.props.onFinished) this.props.onFinished();
     },
 
+    onUnhidePreviewClick: function() {
+        if (global.localStorage) {
+            // FIXME: factor this out with LinkPreviewWidget
+            // FIXME: somehow propagate this to the EventTile such that it updates itself and realises the link has rematerialised
+            global.localStorage.removeItem("hide_preview_" + this.props.mxEvent.getId());
+        }
+        this.props.mxEvent.widgetHidden = false;
```
Comment (richvdh) [inline src/components/views/rooms/MessageContextMenu.js]: I haven't chased through the code - will this need some sort of forceUpdate? or does the `onFinished` take care of that?

File: src/components/views/elements/ImageView.js (hunk 2 above)
Comment (richvdh) [inline src/components/views/elements/ImageView.js]: this applies to where it is used in `react-sdk` too.

File: src/components/views/rooms/MessageContextMenu.js (hunk 4 above)
Comment (ara4n) [inline src/components/views/rooms/MessageContextMenu.js]: I believe onFinished is sufficient as to hide the context menu the overall event tile has to be rendered

**Comment by ara4n** [comment]
@richvdh PTAL

File: src/components/views/elements/ImageView.js (hunk 1 above)
Comment (richvdh) [inline src/components/views/elements/ImageView.js]: done

File: src/components/views/elements/ImageView.js (hunk 2 above)
Comment (richvdh) [inline src/components/views/elements/ImageView.js]: done

File: src/components/views/elements/ImageView.js (hunk 3 above)
Comment (richvdh) [inline src/components/views/elements/ImageView.js]: dine

Comment (richvdh) [inline src/components/views/elements/ImageView.js]: done also

File: src/components/views/rooms/MessageContextMenu.js (hunk 4 above)
Comment (richvdh) [inline src/components/views/rooms/MessageContextMenu.js]: done

**Comment by richvdh** [comment]
lgtm, modulo talk of rewiring the unhiding.

**Comment by ara4n** [comment]
rewiring unhiding done; @richvdh PTAL

File: src/components/views/rooms/MessageContextMenu.js [hunk 5]
```diff
@@ -104,6 +127,18 @@ module.exports = React.createClass({
             </div>
         );
 
+
+        if (global.localStorage) {
```
Comment (richvdh) [inline src/components/views/rooms/MessageContextMenu.js]: any reason not to use `eventTileOps.isWidgetHidden()` here?

**Comment by richvdh** [comment]
looks good otherwise

File: src/components/views/rooms/MessageContextMenu.js (hunk 5 above)
Comment (ara4n) [inline src/components/views/rooms/MessageContextMenu.js]: nope, thinko; fixed :)

**Pull request merged**